### ISSUE TITLE
Do not use O_NONBLOCK on regular files

### ISSUE THREAD

**Issue by Kleidukos** [haskell/core-libraries-committee#166]
From the original bug [GHC uses O_NONBLOCK on regular files, which has no effect, and blocks the runtime](https://gitlab.haskell.org/ghc/ghc/-/issues/15153)

> GHC is trying to use O_NONBLOCK on regular files, which cannot work and will block when used through unsafe foreign calls like that.

The thread is fairly detailed, and quite interesting to read entirely

Current patch lives at : https://gitlab.haskell.org/ghc/ghc/-/issues/15153

@arybczak has provided benchmarks:

```
unknown@electronics io-test $ cabal run -w ghc-9.4.4 io-test -- --csv baseline.csv
Resolving dependencies...
Build profile: -w ghc-9.4.4 -O1
In order, the following will be built (use -v for more details):
 - io-test-1.0 (exe:io-test) (first run)
Configuring executable 'io-test' for io-test-1.0..
Preprocessing executable 'io-test' for io-test-1.0..
Building executable 'io-test' for io-test-1.0..
[1 of 1] Compiling Main             ( src/Main.hs, /home/unknown/Programowanie/haskell/io-test/dist-newstyle/build/x86_64-linux/ghc-9.4.4/io-test-1.0/x/io-test/build/io-test/io-test-tmp/Main.o )
[2 of 2] Linking /home/unknown/Programowanie/haskell/io-test/dist-newstyle/build/x86_64-linux/ghc-9.4.4/io-test-1.0/x/io-test/build/io-test/io-test
All
  read
    2048:  OK (1.16s)
      386  ms ±  19 ms
    4096:  OK (1.09s)
      363  ms ± 5.6 ms
    8192:  OK (1.07s)
      355  ms ±  29 ms
    16384: OK (0.93s)
      308  ms ± 3.4 ms
    32768: OK (0.89s)
      295  ms ± 3.2 ms
    65536: OK (0.86s)
      288  ms ± 4.3 ms

All 6 tests passed (6.01s)
unknown@electronics io-test $ cabal run io-test -- --baseline baseline.csv 
Resolving dependencies...
Build profile: -w ghc-9.4.4.20230216 -O1
In order, the following will be built (use -v for more details):
 - io-test-1.0 (exe:io-test) (first run)
Configuring executable 'io-test' for io-test-1.0..
Preprocessing executable 'io-test' for io-test-1.0..
Building executable 'io-test' for io-test-1.0..
[1 of 1] Compiling Main             ( src/Main.hs, /home/unknown/Programowanie/haskell/io-test/dist-newstyle/build/x86_64-linux/ghc-9.4.4.20230216/io-test-1.0/x/io-test/build/io-test/io-test-tmp/Main.o )
[2 of 2] Linking /home/unknown/Programowanie/haskell/io-test/dist-newstyle/build/x86_64-linux/ghc-9.4.4.20230216/io-test-1.0/x/io-test/build/io-test/io-test
All
  read
    2048:  OK (1.33s)
      442  ms ± 9.7 ms, 14% more than baseline
    4096:  OK (1.26s)
      419  ms ± 5.1 ms, 15% more than baseline
    8192:  OK (1.21s)
      403  ms ±  14 ms, 13% more than baseline
    16384: OK (1.00s)
      332  ms ± 3.3 ms,  7% more than baseline
    32768: OK (0.91s)
      304  ms ± 3.7 ms,  3% more than baseline
    65536: OK (0.87s)
      291  ms ± 5.4 ms,       same as baseline

All 6 tests passed (6.58s)
```

with the following comment:

> My opinion is that it doesn't matter how fast the current code is if it's incorrect. The assumption that file reads are always fast is clearly incorrect, especially nowadays where a lot of code runs on third party instances with restricted amount of storage IOPS like aws.
Anyone who does this has the potential to run into serious trouble that will be almost impossible to debug.
EDIT: also, this isn't a theoretical concern. I've run into this problem, struggled with fixing the root cause for 2 weeks as suddenly nothing made sense anymore (because logging within the application also became unreliable), then an unrelated fix that reduced the amount of IO activity made the problem go away and I stumbled onto [#15153](https://github.com/ghc/ghc/-/issues/15153) randomly while browsing the issue tracker (and only because [@nh2](https://github.com/nh2) refreshed the ticket by posting a MR) a couple months later.

---

From a personal standpoint, I can attest that in industrial environments that make use of network-based storage like AWS, this is a very important fix (disclaimer: All the companies I've worked at in the recent years use such a type of storage for applicative servers).

It must be noted that this is happening in parallel of a [rewrite of the I/O Manager with io_uring](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8073), so while this is a fix for a bug, it is not bound to be a permanent solution.

**Comment by mixphix** [comment]
The actual merge request is located [here](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7986). It's a one-line change that adds a check to see if the file is a `RegularFile`. This seems like a strict improvement to me, but I'd like to hear other members' thoughts!

**Comment by mixphix** [comment]
I am a little concerned about the interplay with the rewrite of the I/O manager. @bgamari could you explain what the intent and effect that will have on the `base` library, if any?

**Comment by bgamari** [comment]
@mixphix, I suspect that  you mean the `io-uring` backend in particular. Using `O_NONBLOCK` in the existing I/O paths should not jeopardize the ability to introduce an `io-uring` backend. Moreover, such a backend would have the advantage of not suffering from the overhead introduced by `O_NONBLOCK`.

**Comment by mixphix** [comment]
Yes, thanks. That's reass"uring". :)

**Comment by Bodigrim** [comment]
As someone who uses Haskell to access slow network drives, I'm utterly in favor.

**Comment by hasufell** [comment]
I'd like to get input from @Rufflewind and @Mistuke 

Note that other packages such as file-io/unix etc may need patching. Otherwise we create subtle differences across the ecosystem. 

E.g.: https://github.com/hasufell/file-io/blob/master/System/File/Posix.hs

**Comment by treeowl** [comment]
Could someone explain the problem with `O_NONBLOCK` and slow network drives?

**Comment by Kleidukos** [comment]
@treeowl it is extensively explained in the ticket on Gitlab.

**Comment by Rufflewind** [comment]
I don't think [turning off O_NONBLOCK](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7986) is the proper solution.

I think the real bug is on [this line](https://hackage.haskell.org/package/base-4.11.1.0/docs/src/GHC.IO.FD.html#readRawBufferPtr):

```hs
  | isNonBlocking fd = unsafe_read
```

It's probably a bad idea to assume that `O_NONBLOCK` is meaningful. The [POSIX spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/open.html#:~:text=errno%20to%20%5BELOOP%5D.-,O_NONBLOCK,-When%20opening%20a) only defines the semantics of `O_NONBLOCK` on certain kinds of files. Implementations like Linux seem to treat it as a hint, honored where possible and ignored if not.

I think the proper fix should be to remove this hack from `readRawBufferPtr` and introduce an alternate (opt-in) API for the rare case where the user wants to avoid the `unsafe` `foreign` overhead.

---

On a separate matter, it might be worth also diving into the `unix` library and make sure that blockable `foreign` calls are marked as `safe`. For example, I noticed [`openAt`](https://github.com/ghc/packages-unix/blob/31eb772051a2ef4057652a169c2688deed735d9b/System/Posix/IO/Common.hsc#L246) is unsafe, but I believe it's possible for `openAt` to block.

**Comment by Ericson2314** [comment]
> I am a little concerned about the interplay with the rewrite of the I/O manager.

Yeah architecturally, I few this as an implementation detail of the I/O manager, which may be in `base` not `rts` but is (to me) still morally part of the runtime.

This makes it a funny thing to consider here, procedurally.

------

It is possible there could be some "pre IO manager" IO primitives that are more directly exposing what the OS provides, unvarnished. These would be very dangerous to use in most code, but also great building blocks for writing the IO manager in the first place.

That strikes me as a good architecture, and then those "pre IO manager" IO primitives would be squarely in CLC remit since they are regular public library functions, with little implementation wiggle room left by their "public" behavior.

**Comment by Bodigrim** [comment]
> @treeowl it is extensively explained in the ticket's thread on Gitlab.

@Kleidukos @arybczak @nh2 could you please provide a self-contained summary, which can serve as a proposal? Precisely because GitLab tickets are extremely extensive. It's not like an average Haskell developer has any idea about `O_NONBLOCK` at all. Please include reflection on suggestions from https://github.com/haskell/core-libraries-committee/issues/166#issuecomment-1548618063.

**Comment by nh2** [comment]
@Bodigrim Yes, excellent summary!

Some detail remarks:

> While regular files cannot block indefinitely as other types of file descriptors such as sockets and pipes can, they always block the caller for a brief amount of time: you cannot read a file from a floppy drive or network share without delay

Where "brief" can in practice, and today's real-world industry settings be

* 10 milliseconds for a spinning hard drive seek, or
* 2 seconds for a spinning hard drive that's under load from other requests, or
* 2 seconds for a network share far away, or
* 20 hours for a network share currently undergoing maintenance or having a network disconnect (file systems such as Ceph, sshfs, or other FUSE based network file systems block opening files when the connection is down),

Hit any of those for multiple files concurrently, and the Haskell program freezes for a while. Thus motivating concretely why this proposal exists.

**Comment by Bodigrim** [comment]
I put up a new, slightly extended MR https://gitlab.haskell.org/ghc/ghc/-/merge_requests/11338. Besides some pieces of documentation, the only material change is one line:
```diff
-                fdIsNonBlocking = fromEnum is_nonblock
+                fdIsNonBlocking = fromEnum (is_nonblock && fd_type /= RegularFile && fd_type /= RawDevice)
```
For anyone new to this discussion, the summary of the issue is available above, at https://github.com/haskell/core-libraries-committee/issues/166#issuecomment-1597746025.

I'll change hats and trigger a vote shortly.

**Comment by Bodigrim** [comment]
Actually, given that this is a non-trivial piece of code and it would be unfortunate to spoil multiyear efforts by calling a vote too early, let me ask for non-binding opinions first.

Dear CLC members, could you please read https://github.com/haskell/core-libraries-committee/issues/166#issuecomment-1597746025 and opine on https://gitlab.haskell.org/ghc/ghc/-/merge_requests/11338? 
@tomjaguarpaw @parsonsmatt @angerman @velveteer @hasufell @mixphix

**Comment by angerman** [comment]
@Bodigrim thank you for the excellent summary! I currently have trouble finding a common case where this would break existing code and behaviour, and other than potentially reducing stalls, I can't find any. As such I'm in favour of this change.
+1

**Comment by velveteer** [comment]
The performance regression is somewhat concerning, no? Is Marlow's comment (https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7986#note_502536) being considered in this proposal as well?

Otherwise, I am in favor of this.

**Comment by Bodigrim** [comment]
> The performance regression is somewhat concerning, no?

We are essentially fixing a DDoS vulnerability (= a single slow file I/O can render a multithreaded program unresponsible), so a modest performance regression is acceptable. We'd better get it correct first, optimize constant factor later.

> Is Marlow's comment (https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7986#note_502536) being considered in this proposal as well?

My view is that such optimization could be a subject of a future proposal. It's not straightforward to implement: at least one has to extend `data FD` with a new field keeping type of a descriptor (pipe / socket or regular file). This itself may trigger non-trivial performance consequences and further investigation. I don't really have an expertise in this area; I just stepped up to refresh the MR because original proposers were seemingly inactive.

**Comment by Rufflewind** [comment]
> at least one has to extend `data FD` with a new field keeping type of a descriptor (pipe / socket or regular file).

If we ever go with that approach, I would recommend adding a `SafeOrUnsafe` field rather than a `FileType` field. I do not think it was ever a good idea to try and guess `SafeOrUnsafe` from `FileType` to begin with. I think the users should make that choice explicitly based on their use case, with `Safe` always a safe albeit slow default.

**Comment by Bodigrim** [comment]
Any more non-binding opinions? @tomjaguarpaw @mixphix @parsonsmatt @hasufell

**Comment by hasufell** [comment]
I'm +1

The performance regression can be fixed later. I'm kind of baffled by the current behavior and wasn't aware of it.

----

As a note: the title is confusing. This has technically nothing to do with passing O_NONBLOCK to `open`, but whether we use `safe` or `unsafe` call to `read`, which is about how Haskell deals with FFI. And  the patch does not remove the O_NONBLOCK flags from `open`. They're still passed to the FFI functions.

**Comment by hasufell** [comment]
> A common pattern is to define both safe and unsafe FFI bindings and switch between them depending on the size of payload or RTS configuration. Here is an example from [bytestring](https://hackage.haskell.org/package/bytestring-0.11.4.0/docs/src/Data.ByteString.Short.Internal.html#isValidUtf8)

RTS configuration for this sounds like a good way to allow people to opt out of the changed behavior or tweak it?

**Comment by nh2** [comment]
> the title is confusing. This has technically nothing to do with passing O_NONBLOCK to `open`, but whether we use `safe` or `unsafe` call to `read`

@hasufell You are right. That's my fault: I orignally wrote the commit message `Do not use O_NONBLOCK on regular files`, when in fact the giving of `O_NONBLOCK` is unchanged. The message should be fixed as you suggest.

**Comment by Bodigrim** [comment]
Dear CLC members, let's vote on https://gitlab.haskell.org/ghc/ghc/-/merge_requests/11338/diffs to fix a DDoS vulnerability (= a single slow file I/O, e. g., network shared drive, can render an entire multithreaded program unresponsive). This patch is backwards-compatible and does not introduce new API.

@hasufell @tomjaguarpaw @parsonsmatt @mixphix @velveteer @angerman 

----

+1 from me.

**Comment by angerman** [comment]
+1, as it doesn't break existing code. I see the concern for potential change in runtime behavior. That does however rather lead me to believe we should have an additional exploit api to let the end user chose rather than rely on a heuristic.

**Comment by hasufell** [comment]
+1

**Comment by velveteer** [comment]
+1

**Comment by mixphix** [comment]
+1

**Comment by Bodigrim** [comment]
Thanks all, that’s enough votes to approve unconditionally. 

Indeed this area may benefit from a more invasive contribution, refactoring `data FD` along the lines of the discussion above (+ impact assessment, + migration policy). An enthusiastic contributor is welcome :)

**Comment by hasufell** [comment]
This caused a regression in hinotify: https://gitlab.haskell.org/ghc/ghc/-/issues/25199
